File: nexo-desktop/src/main/descubrimiento.js
'use strict';

// Descubrimiento por WiFi con Bonjour (mDNS). El PC se anuncia en la red local
// como servicio "_nexo._tcp"; la app del iPhone lo encuentra en una lista y se
// conecta sin que nadie teclee una direccion IP. Es la ruta alternativa al
// cable: mismo protocolo por encima, distinto camino.

const os = require('os');
const { Bonjour } = require('bonjour-service');

const TIPO = 'nexo';

class Anunciante {
  constructor() {
    this.bonjour = new Bonjour();
    this.servicio = null;
  }

  // Publica el PC como receptor Nexo en el puerto del transporte WiFi.
  anunciar(puerto) {
    this.detener();
    this.servicio = this.bonjour.publish({
      name: `Nexo en ${os.hostname()}`,
      type: TIPO,
      port: puerto,
      txt: { version: '1', app: 'Nexo Desktop', equipo: os.hostname() },
    });
    return this.servicio;
  }

  detener() {
    if (this.servicio) {
      try {
        this.servicio.stop();
      } catch {
        /* ya detenido */
      }
      this.servicio = null;
    }
  }

  cerrar() {
    this.detener();
    try {
      this.bonjour.destroy();
    } catch {
      /* ya cerrado */
    }
  }
}

// Busca otros PCs con Nexo en la red (util para pruebas y para la app iOS, que
// hace lo mismo desde Swift). onCambio recibe la lista actual.
class Buscador {
  constructor(onCambio) {
    this.bonjour = new Bonjour();
    this.onCambio = onCambio || (() => {});
    this.encontrados = new Map();
    this.navegador = null;
  }

  iniciar() {
    this.navegador = this.bonjour.find({ type: TIPO }, () => {});
    const emitir = () => this.onCambio([...this.encontrados.values()]);

    this.navegador.on('up', (s) => {
      this.encontrados.set(s.fqdn || s.name, {
        nombre: s.name,
        puerto: s.port,
        direcciones: s.addresses || [],
        equipo: s.txt?.equipo || null,
      });
      emitir();
    });
    this.navegador.on('down', (s) => {
      this.encontrados.delete(s.fqdn || s.name);
      emitir();
    });
  }

  cerrar() {
    try {
      this.navegador?.stop();
      this.bonjour.destroy();
    } catch {
      /* ya cerrado */
    }
  }
}

module.exports = { Anunciante, Buscador, TIPO };

File: nexo-ios/Sources/Descubrimiento.swift
import Foundation
import Network

// Busca PCs con Nexo en la red WiFi (servicio _nexo._tcp que anuncia
// descubrimiento.js en el PC). Se usa solo para la ruta WiFi; por cable no hace
// falta descubrir nada. Con NWBrowser, la API moderna de Network.

struct PCNexo: Identifiable, Equatable {
    let id: String       // nombre del servicio
    let nombre: String
    let endpoint: NWEndpoint
}

final class BuscadorPC {
    private var browser: NWBrowser?
    var alCambio: (([PCNexo]) -> Void)?
    private(set) var encontrados: [PCNexo] = []

    func iniciar() {
        let params = NWParameters()
        params.includePeerToPeer = true
        let browser = NWBrowser(for: .bonjour(type: "_nexo._tcp", domain: nil), using: params)

        browser.browseResultsChangedHandler = { [weak self] resultados, _ in
            guard let self = self else { return }
            self.encontrados = resultados.compactMap { r in
                if case let .service(name, _, _, _) = r.endpoint {
                    return PCNexo(id: name, nombre: name, endpoint: r.endpoint)
                }
                return nil
            }
            self.alCambio?(self.encontrados)
        }
        browser.start(queue: .main)
        self.browser = browser
    }

    func detener() {
        browser?.cancel()
        browser = nil
    }
}
